File: injectai/console.py
"""The meta console: the operator's home prompt.

This is the layer the operator lands in before/after engagements. It exposes the
engagement catalogue, the training modules, settings and progress. Selecting an
engagement drops into the live :class:`OperationalShell`.
"""
from __future__ import annotations

from . import ui
from .game import GameState, save_config
from .shell import LeaveShell, OperationalShell
from .training import run_training
from .world import all_scenarios


class Console:
    def __init__(self, state: GameState):
        self.state = state
        self.scenarios = all_scenarios()
        self.commands = {
            "help": self._help, "?": self._help,
            "scenarios": self._scenarios, "targets": self._scenarios,
            "engagements": self._scenarios, "ls": self._scenarios,
            "brief": self._brief, "info": self._brief,
            "engage": self._engage, "connect": self._engage, "start": self._engage,
            "training": self._training, "train": self._training, "learn": self._training,
            "settings": self._settings, "config": self._settings,
            "status": self._status, "progress": self._status,
            "whoami": self._whoami, "clear": self._clear,
            "exit": self._quit, "quit": self._quit, "logout": self._quit,
        }

    # -- prompt -----------------------------------------------------------
    def prompt(self) -> str:
        op = self.state.operator
        return (ui.c(op, "brightcyan", "bold") + ui.c("@", "grey")
                + ui.c("injectai", "cyan") + ui.c(" > ", "grey"))

    # -- main loop --------------------------------------------------------
    def loop(self) -> None:
        while True:
            try:
                raw = input(self.prompt())
            except EOFError:
                ui.write()
                return
            except KeyboardInterrupt:
                ui.write("\n" + ui.c("(type 'exit' to quit)", "grey"))
                continue
            line = raw.strip()
            if not line:
                continue
            parts = line.split()
            fn = self.commands.get(parts[0].lower())
            if not fn:
                ui.write(ui.c(f"{parts[0]}: unknown command. Type `help`.", "red"))
                continue
            try:
                if fn(parts[1:]) == "QUIT":
                    return
            except LeaveShell:
                pass

    # -- commands ---------------------------------------------------------
    def _help(self, args):
        ui.write()
        rows = [
            ("scenarios", "list available engagements / targets"),
            ("brief <code>", "read the briefing for an engagement"),
            ("engage <code>", "deploy into the engagement's operator shell"),
            ("training [tier]", "guided modules (easy | medium | hard)"),
            ("status", "your progress and recovered objectives"),
            ("settings", "display & interaction preferences"),
            ("clear", "clear the screen"),
            ("exit", "close the console"),
        ]
        for k, v in rows:
            ui.write("  " + ui.c(k.ljust(16), "cyan", "bold") + ui.c(v, "white"))
        ui.write()

    def _scenarios(self, args):
        ui.write()
        ui.write(ui.c("  AVAILABLE ENGAGEMENTS", "cyan", "bold"))
        ui.hr(60)
        completed = self.state.config.get("completed", [])
        for code, net in self.scenarios.items():
            done = ui.c(" [resolved]", "brightgreen") if code in completed else ""
            ui.write("  " + ui.c(code.ljust(12), "brightyellow", "bold")
                     + ui.c(net.difficulty.ljust(24), "grey") + done)
            ui.write("    " + ui.c(net.title, "white"))
            ui.write("    " + ui.c(net.summary, "grey"))
            ui.write()
        ui.write(ui.c("  brief <code> for details, engage <code> to deploy.", "grey"))
        ui.write()

    def _brief(self, args):
        if not args:
            ui.write(ui.c("usage: brief <code>  (see `scenarios`)", "red"))
            return
        net = self.scenarios.get(args[0].upper())
        if not net:
            ui.write(ui.c(f"no engagement '{args[0]}'.", "red"))
            return
        ui.write()
        ui.box(f"{net.code}  ::  {net.difficulty}", [], width=66)
        ui.write(ui.c(net.title, "brightyellow", "bold"))
        ui.write()
        for ln in net.brief.splitlines():
            ui.write("  " + ln)
        ui.write()
        ui.write(ui.c(f"  engage {net.code}   to begin.", "cyan"))
        ui.write()

    def _engage(self, args):
        if not args:
            ui.write(ui.c("usage: engage <code>  (see `scenarios`)", "red"))
            return
        net = self.scenarios.get(args[0].upper())
        if not net:
            ui.write(ui.c(f"no engagement '{args[0]}'.", "red"))
            return
        self.state.load_network(net)
        self._enter_shell(net)

    def _enter_shell(self, net):
        ui.clear()
        ui.box("ENGAGEMENT ACTIVE  ::  " + net.code, [
            ui.c(net.title, "white"),
            ui.c("Scope: " + net.subnet, "grey"),
            "",
            ui.c("Your operator workstation is " + self.state.operator_host.ip + ".", "grey"),
            ui.c("`help` lists tooling, `status` tracks progress, `back` exits.", "grey"),
        ], width=66)
        ui.write()
        if self.state.config.get("hints"):
            ui.write(ui.c("  starting point: in-scope host(s) are reachable; begin "
                          "with recon (nmap).", "grey"))
            ui.write()
        shell = OperationalShell(self.state)
        while True:
            try:
                raw = input(shell.prompt())
            except EOFError:
                ui.write()
                return
            except KeyboardInterrupt:
                ui.write("\n" + ui.c("(type 'back' to return to the console)", "grey"))
                continue
            try:
                shell.run_line(raw)
            except LeaveShell:
                ui.write(ui.c("disengaged. returning to console.", "grey"))
                return

    def _training(self, args):
        tier = args[0].lower() if args else None
        if tier not in ("easy", "medium", "hard"):
            ui.write()
            ui.write(ui.c("  TRAINING MODULES", "cyan", "bold"))
            ui.hr(48)
            ui.write("  " + ui.c("easy".ljust(8), "brightgreen", "bold")
                     + ui.c("full command walkthrough", "white"))
            ui.write("  " + ui.c("medium".ljust(8), "brightyellow", "bold")
                     + ui.c("conceptual hints only", "white"))
            ui.write("  " + ui.c("hard".ljust(8), "brightred", "bold")
                     + ui.c("objectives only -- no guidance", "white"))
            ui.write()
            try:
                tier = input(ui.c("  select tier (easy/medium/hard): ", "cyan")).strip().lower()
            except (EOFError, KeyboardInterrupt):
                ui.write()
                return
            if tier not in ("easy", "medium", "hard"):
                ui.write(ui.c("  cancelled.", "grey"))
                return
        run_training(self.state, tier)

    def _settings(self, args):
        cfg = self.state.config
        if args and len(args) >= 2:
            key, val = args[0].lower(), args[1].lower()
            self._set(key, val)
            return
        ui.write()
        ui.write(ui.c("  SETTINGS", "cyan", "bold"))
        ui.hr(48)
        items = [
            ("colour", cfg.get("colour"), "ANSI colour output"),
            ("typing", cfg.get("typing"), "per-character typing effect"),
            ("speed", cfg.get("speed"), "effect/animation speed multiplier"),
            ("hints", cfg.get("hints"), "inline tactical hints"),
            ("operator", cfg.get("operator"), "operator handle (prompt name)"),
        ]
        for k, v, desc in items:
            ui.write("  " + ui.c(k.ljust(10), "brightyellow")
                     + ui.c(str(v).ljust(10), "white") + ui.c(desc, "grey"))
        ui.write()
        ui.write(ui.c("  change with: settings <key> <value>", "grey"))
        ui.write(ui.c("  e.g. settings typing on   |   settings operator ghost", "grey"))
        ui.write()

    def _set(self, key, val):
        cfg = self.state.config
        bools = {"on": True, "off": False, "true": True, "false": False,
                 "yes": True, "no": False, "1": True, "0": False}
        if key in ("colour", "color", "typing", "hints"):
            k = "colour" if key in ("colour", "color") else key
            if val not in bools:
                ui.write(ui.c("  value must be on/off.", "red")); return
            cfg[k] = bools[val]
        elif key == "speed":
            try:
                cfg["speed"] = max(0.1, float(val))
            except ValueError:
                ui.write(ui.c("  speed must be a number.", "red")); return
        elif key == "operator":
            from .game import build_operator_host
            cfg["operator"] = val
            self.state.operator = val
            self.state.operator_host = build_operator_host(val)
        else:
            ui.write(ui.c(f"  unknown setting '{key}'.", "red")); return
        save_config(cfg)
        # re-apply presentation prefs live
        ui.set_style(ui.Style(cfg.get("colour", True), cfg.get("typing", False),
                              cfg.get("speed", 1.0)))
        ui.write(ui.c(f"  {key} = {val}", "brightgreen"))

    def _status(self, args):
        cfg = self.state.config
        completed = cfg.get("completed", [])
        ui.write()
        ui.box("OPERATOR STATUS", [
            ui.c("handle:        " + self.state.operator, "white"),
            ui.c("training best: " + (cfg.get("training_best") or "none"), "white"),
            ui.c("engagements resolved: %d / %d" % (len(completed), len(self.scenarios)), "white"),
        ], width=60)
        for code, net in self.scenarios.items():
            mark = ui.c("✓", "brightgreen") if code in completed else ui.c("·", "grey")
            ui.write(f"   {mark} {code.ljust(12)} {ui.c(net.title, 'grey')}")
        ui.write()

    def _whoami(self, args):
        ui.write(self.state.operator + "@injectai")

    def _clear(self, args):
        ui.clear()

    def _quit(self, args):
        ui.write(ui.c("session closed.", "grey"))
        return "QUIT"

File: console.js
/* injectAI web build -- meta console, training driver and boot.
 * Browser port of console.py / training.py / app.py. */
"use strict";

(function () {
  const { Shell, GameState, loadConfig, saveConfig } = window.SHELL;

  const BANNER =
    "   _      _           _      _    ___\n" +
    "  (_)__  (_)__ ____  / /_   / |  /  _/\n" +
    " / / _ \\/ / -_) __/ /_  _/ /  _|_/ /\n" +
    "/_/_//_/_/\\__/\\__/   /_/  /_/ |_/___/";

  let state, scenarios;

  /* ---- training -------------------------------------------------- */
  const TARGET = "10.13.37.20";
  const hasCred = (st, user) => st.loot.creds.some((x) => x[1] === user);
  const hasSession = (st, ip, user) => st.sessions.some((s) => s.host.ip === ip && s.user === user);

  function tasks() {
    return [
      { obj: "Identify which services the host exposes and their versions.",
        chk: (st) => (st.scanned[TARGET] || new Set()).has(22),
        cmd: "nmap -sV " + TARGET,
        hint: "Fingerprint the host with a version scan before anything else." },
      { obj: "Recover a valid login for the 'student' service account.",
        chk: (st) => hasCred(st, "student"),
        cmd: "hydra -l student -P wordlists/common.txt ssh://" + TARGET,
        hint: "The account uses a weak seasonal password. An online dictionary attack against SSH will find it." },
      { obj: "Use the recovered credential to obtain an interactive foothold.",
        chk: (st) => hasSession(st, TARGET, "student"),
        cmd: "ssh student@" + TARGET + "      (password from the previous step)",
        hint: "Log in over SSH as the account whose password you just cracked." },
      { obj: "Enumerate the account's sudo rights and abuse them to reach root.",
        chk: (st) => hasSession(st, TARGET, "root"),
        cmd: "sudo -l         then:    sudo find . -exec /bin/bash \\; -quit",
        hint: "Check `sudo -l`. A whitelisted binary that can run other programs (GTFOBins) is a direct path to a root shell." },
      { obj: "Recover the proof artefact from the root account.",
        chk: (st) => st.config.completed.includes("TRN-SANDBOX"),
        cmd: "cat /root/flag.txt",
        hint: "Read root's proof file now that you are uid 0." },
    ];
  }

  function guidance(task, diff) {
    if (diff === "easy") C.write(C.c("    run: ", "grey") + C.c(task.cmd, "brightyellow"));
    else if (diff === "medium") C.write(C.c("    hint: " + task.hint, "grey"));
  }

  async function runTraining(diff) {
    diff = diff.toLowerCase();
    const tier = { easy: "Easy / Recruit", medium: "Medium / Operator", hard: "Hard / Specialist" }[diff] || "Easy / Recruit";
    state.loadNetwork(window.WORLD.allScenarios()["TRN-SANDBOX"]);
    const sh = new Shell(state);
    const T = tasks();
    let idx = 0;
    C.clear();
    C.box("TRAINING MODULE  ::  " + tier, [
      C.c("Lab host: TRN-SANDBOX-01 (10.13.37.20)", "white"),
      C.c("This is the live operator shell. Tooling is identical to a real", "grey"),
      C.c("engagement; objectives below track your progress.", "grey"),
      "",
      C.c("Type `objective` to repeat the current goal, `back` to leave.", "grey"),
    ], 66);
    C.write("");
    const showObjective = () => {
      if (idx >= T.length) return;
      C.write(C.c("  OBJECTIVE " + (idx + 1) + "/" + T.length + ": ", "cyan", "bold") + C.c(T[idx].obj, "white"));
      guidance(T[idx], diff);
      C.write("");
    };
    showObjective();
    while (true) {
      if (idx >= T.length) {
        C.write("");
        C.box("MODULE COMPLETE", [
          C.c("All objectives met. You chained service discovery, a", "white"),
          C.c("credential attack, a foothold and a sudo/GTFOBins escalation", "white"),
          C.c("into full root access -- a complete real-world kill chain.", "white"),
        ], 66, "green");
        const order = { easy: 1, medium: 2, hard: 3 };
        if ((order[diff] || 0) >= (order[state.config.training_best] || 0)) { state.config.training_best = diff; saveConfig(state.config); }
        C.write("");
        return;
      }
      const raw = await C.readLine(sh.prompt());
      const cmd = raw.trim();
      if (["objective", "objectives", "goal"].includes(cmd)) { showObjective(); continue; }
      if (cmd === "hint") { C.write(C.c("    hint: " + T[idx].hint, "grey")); continue; }
      const sig = await sh.runLine(raw);
      if (sig === "BACK") { C.write(C.c("leaving training module.", "grey")); return; }
      let advanced = false;
      while (idx < T.length && T[idx].chk(state)) {
        C.write("");
        C.write(C.c("  [✓] objective " + (idx + 1) + " complete: ", "brightgreen", "bold") + C.c(T[idx].obj, "green"));
        idx++; advanced = true;
      }
      if (advanced && idx < T.length) { C.write(""); showObjective(); }
    }
  }

  /* ---- engagement shell ----------------------------------------- */
  async function enterShell(net) {
    state.loadNetwork(net);
    C.clear();
    C.box("ENGAGEMENT ACTIVE  ::  " + net.code, [
      C.c(net.title, "white"),
      C.c("Scope: " + net.subnet, "grey"),
      "",
      C.c("Your operator workstation is " + state.operatorHost.ip + ".", "grey"),
      C.c("`help` lists tooling, `status` tracks progress, `back` exits.", "grey"),
    ], 66);
    C.write("");
    if (state.config.hints) {
      C.write(C.c("  starting point: in-scope host(s) are reachable; begin with recon (nmap).", "grey"));
      C.write("");
    }
    const sh = new Shell(state);
    while (true) {
      const raw = await C.readLine(sh.prompt());
      const sig = await sh.runLine(raw);
      if (sig === "BACK") { C.write(C.c("disengaged. returning to console.", "grey")); return; }
    }
  }

  /* ---- meta console --------------------------------------------- */
  const CONSOLE = {};
  CONSOLE.help = function () {
    C.write("");
    [["scenarios", "list available engagements / targets"],
    ["brief <code>", "read the briefing for an engagement"],
    ["engage <code>", "deploy into the engagement's operator shell"],
    ["training [tier]", "guided modules (easy | medium | hard)"],
    ["status", "your progress and recovered objectives"],
    ["settings", "display & interaction preferences"],
    ["clear", "clear the screen"],
    ].forEach(([k, v]) => C.write("  " + C.c(k.padEnd(16), "cyan", "bold") + C.c(v, "white")));
    C.write("");
  };
  CONSOLE.scenarios = function () {
    C.write("");
    C.write(C.c("  AVAILABLE ENGAGEMENTS", "cyan", "bold"));
    C.hr(60);
    const done = state.config.completed;
    Object.keys(scenarios).forEach((code) => {
      const n = scenarios[code];
      const tag = done.includes(code) ? C.c(" [resolved]", "brightgreen") : "";
      C.write("  " + C.c(code.padEnd(12), "brightyellow", "bold") + C.c(n.difficulty.padEnd(24), "grey") + tag);
      C.write("    " + C.c(n.title, "white"));
      C.write("    " + C.c(n.summary, "grey"));
      C.write("");
    });
    C.write(C.c("  brief <code> for details, engage <code> to deploy.", "grey"));
    C.write("");
  };
  CONSOLE.brief = function (args) {
    if (!args.length) { C.write(C.c("usage: brief <code>  (see `scenarios`)", "red")); return; }
    const n = scenarios[args[0].toUpperCase()];
    if (!n) { C.write(C.c("no engagement '" + args[0] + "'.", "red")); return; }
    C.write("");
    C.box(n.code + "  ::  " + n.difficulty, [], 66);
    C.write(C.c(n.title, "brightyellow", "bold"));
    C.write("");
    n.brief.split("\n").forEach((ln) => C.write("  " + C.escapeHtml(ln)));
    C.write("");
    C.write(C.c("  engage " + n.code + "   to begin.", "cyan"));
    C.write("");
  };
  CONSOLE.engage = async function (args) {
    if (!args.length) { C.write(C.c("usage: engage <code>  (see `scenarios`)", "red")); return; }
    const n = scenarios[args[0].toUpperCase()];
    if (!n) { C.write(C.c("no engagement '" + args[0] + "'.", "red")); return; }
    await enterShell(n);
  };
  CONSOLE.training = async function (args) {
    let tier = args.length ? args[0].toLowerCase() : null;
    if (!["easy", "medium", "hard"].includes(tier)) {
      C.write("");
      C.write(C.c("  TRAINING MODULES", "cyan", "bold"));
      C.hr(48);
      C.write("  " + C.c("easy".padEnd(8), "brightgreen", "bold") + C.c("full command walkthrough", "white"));
      C.write("  " + C.c("medium".padEnd(8), "brightyellow", "bold") + C.c("conceptual hints only", "white"));
      C.write("  " + C.c("hard".padEnd(8), "brightred", "bold") + C.c("objectives only -- no guidance", "white"));
      C.write("");
      tier = (await C.readLine(C.c("  select tier (easy/medium/hard): ", "cyan"))).trim().toLowerCase();
      if (!["easy", "medium", "hard"].includes(tier)) { C.write(C.c("  cancelled.", "grey")); return; }
    }
    await runTraining(tier);
  };
  CONSOLE.settings = function (args) {
    const cfg = state.config;
    if (args.length >= 2) { setSetting(args[0].toLowerCase(), args[1].toLowerCase()); return; }
    C.write("");
    C.write(C.c("  SETTINGS", "cyan", "bold"));
    C.hr(48);
    [["colour", cfg.colour, "ANSI colour output"],
    ["hints", cfg.hints, "inline tactical hints"],
    ["operator", cfg.operator, "operator handle (prompt name)"],
    ].forEach(([k, v, d]) => C.write("  " + C.c(String(k).padEnd(10), "brightyellow") + C.c(String(v).padEnd(10), "white") + C.c(d, "grey")));
    C.write("");
    C.write(C.c("  change with: settings <key> <value>", "grey"));
    C.write(C.c("  e.g. settings hints off   |   settings operator ghost", "grey"));
    C.write("");
  };
  function setSetting(key, val) {
    const cfg = state.config;
    const bools = { on: true, off: false, true: true, false: false, yes: true, no: false, "1": true, "0": false };
    if (["colour", "color", "hints"].includes(key)) {
      const k = key === "color" ? "colour" : key;
      if (!(val in bools)) { C.write(C.c("  value must be on/off.", "red")); return; }
      cfg[k] = bools[val];
      if (k === "colour") document.body.classList.toggle("nocolour", !cfg.colour);
    } else if (key === "operator") {
      cfg.operator = val; state.operator = val;
      state.operatorHost = window.SHELL.buildOperatorHost(val);
      state.sessions = [{ host: state.operatorHost, user: val, cwd: state.homeNode(state.operatorHost, val), label: "local" }];
    } else { C.write(C.c("  unknown setting '" + key + "'.", "red")); return; }
    saveConfig(cfg);
    C.write(C.c("  " + key + " = " + val, "brightgreen"));
  }
  CONSOLE.status = function () {
    const cfg = state.config;
    C.write("");
    C.box("OPERATOR STATUS", [
      C.c("handle:        " + state.operator, "white"),
      C.c("training best: " + (cfg.training_best || "none"), "white"),
      C.c("engagements resolved: " + cfg.completed.length + " / " + Object.keys(scenarios).length, "white"),
    ], 60);
    Object.keys(scenarios).forEach((code) => {
      const mark = cfg.completed.includes(code) ? C.c("✓", "brightgreen") : C.c("·", "grey");
      C.write("   " + mark + " " + code.padEnd(12) + " " + C.c(scenarios[code].title, "grey"));
    });
    C.write("");
  };
  CONSOLE.whoami = function () { C.write(state.operator + "@injectai"); };
  CONSOLE.clear = function () { C.clear(); };

  const ALIASES = {
    targets: "scenarios", engagements: "scenarios", ls: "scenarios",
    info: "brief", connect: "engage", start: "engage",
    train: "training", learn: "training", config: "settings", progress: "status", "?": "help",
  };

  async function consoleLoop() {
    while (true) {
      const op = state.operator;
      const prompt = C.c(op, "brightcyan", "bold") + C.c("@", "grey") + C.c("injectai", "cyan") + C.c(" > ", "grey");
      const raw = await C.readLine(prompt);
      const line = raw.trim();
      if (!line) continue;
      const parts = line.split(/\s+/);
      let name = parts[0].toLowerCase();
      if (name === "exit" || name === "quit" || name === "logout") {
        C.write(C.c("session closed. (refresh the page to start a new session)", "grey")); return;
      }
      name = ALIASES[name] || name;
      const fn = CONSOLE[name];
      if (!fn) { C.write(C.c(parts[0] + ": unknown command. Type `help`.", "red")); continue; }
      await fn(parts.slice(1));
    }
  }

  /* ---- boot ------------------------------------------------------ */
  async function boot() {
    const cfg = loadConfig();
    document.body.classList.toggle("nocolour", !cfg.colour);
    state = new GameState(cfg);
    scenarios = window.WORLD.allScenarios();
    C.clear();
    C.write(C.c(BANNER, "brightcyan"));
    C.write(C.c("  injectAI  ", "brightcyan", "bold") + C.c("offensive security operations console", "grey"));
    C.write(C.c("  ────────────────────────────────────────────────", "grey"));
    const seq = ["loading tooling profiles", "mounting engagement catalogue",
      "initialising virtual network fabric", "operator workstation online"];
    for (const line of seq) {
      C.write(C.c("  [ok] ", "brightgreen") + C.c(line, "grey"));
      await C.sleep(120);
    }
    C.write("");
    C.write(C.c("  Authorised simulation environment. No real systems are", "grey"));
    C.write(C.c("  contacted. Type `help` to begin, `scenarios` to list targets.", "grey"));
    C.write("");
    await consoleLoop();
  }

  // C is the shared IO surface exposed by app.js bootstrap
  const C = window.C;
  window.addEventListener("DOMContentLoaded", boot);
})();
